controllers/keyword_test.go

```diff
@@ -0,0 +1,131 @@
+	db.GetDB = func() *gorm.DB {
+		return database
+	}
+
+	_ = db.GetDB().AutoMigrate(&models.User{}, &models.Keyword{})
+
+	s.engine = testConfig.GetRouter(true)
+	new(LoginController).applyRoutes(EnsureGuestUserGroup(s.engine))
+	new(KeywordController).applyRoutes(EnsureAuthenticatedUserGroup(s.engine))
+
+	email := faker.Email()
+	password := faker.Password()
+
+	hashedPassword, _ := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
+	user := models.User{Email: email, Password: string(hashedPassword)}
+	db.GetDB().Create(&user)
+
+	formData := url.Values{}
+	formData.Set("email", email)
+	formData.Set("password", password)
+
+	headers := http.Header{}
+	headers.Set("Content-Type", "application/x-www-form-urlencoded")
+	response := testHttp.PerformRequest(s.engine, "POST", "/login", headers, formData)
+
+	s.cookie = response.Header().Get("Set-Cookie")
+}
+
+func (s *KeywordDbTestSuite) TearDownTest() {
+	db.GetDB().Exec("DELETE FROM keywords")
+	db.GetDB().Exec("DELETE FROM users")
+}
+
+func TestKeywordDbTestSuite(t *testing.T) {
+	suite.Run(t, new(KeywordDbTestSuite))
+}
+
+func (s *KeywordDbTestSuite) TestUploadKeywordWithValidParams() {
+	headers, payload := testFile.CreateMultipartPayload("tests/fixture/adword_keywords.csv")
+	headers.Set("Cookie", s.cookie)
+
+	response := testHttp.PerformFileUploadRequest(s.engine, "POST", "/keyword", headers, payload)
+
+	assert.Equal(s.T(), http.StatusFound, response.Code)
+	assert.Equal(s.T(), "/keyword", response.Header().Get("Location"))
+}
+
+func (s *KeywordDbTestSuite) TestUploadKeywordWithBlankPayload() {
+	headers := http.Header{}
+	headers.Set("Cookie", s.cookie)
+
+	response := testHttp.PerformFileUploadRequest(s.engine, "POST", "/keyword", headers, &bytes.Buffer{})
+
+	p, err := ioutil.ReadAll(response.Body)
+	isKeywordPage := err == nil && strings.Index(string(p), "<title>Keyword</title>") > 0
+	pageError := err == nil && strings.Index(string(p), "File is required") > 0
+
+	assert.Equal(s.T(), http.StatusBadRequest, response.Code)
+	assert.Equal(s.T(), true, isKeywordPage)
+	assert.Equal(s.T(), true, pageError)
+}
+
+func TestDisplayKeywordWithGuestUser(t *testing.T) {
+	engine := testConfig.GetRouter(true)
+	new(KeywordController).applyRoutes(EnsureAuthenticatedUserGroup(engine))
+
+	response := testHttp.PerformRequest(engine, "GET", "/keyword", nil, nil)
+
+	assert.Equal(t, http.StatusFound, response.Code)
+	assert.Equal(t, "/login", response.Header().Get("Location"))
+}
+
+func TestUploadKeywordWithGuestUser(t *testing.T) {
+	engine := testConfig.GetRouter(true)
+	new(KeywordController).applyRoutes(EnsureAuthenticatedUserGroup(engine))
+
+	response := testHttp.PerformRequest(engine, "POST", "/keyword", nil, nil)
+
+	assert.Equal(t, http.StatusFound, response.Code)
+	assert.Equal(t, "/login", response.Header().Get("Location"))
+}
+
+func TestDisplayKeywordWithAuthenticatedUser(t *testing.T) {
```

Comment:
Minor thing but shouldn't we group all tests cases for display and upload together. Now it seems messy which makes it hard to follow:

```
upload - TestUploadKeywordWithValidParams
upload - TestUploadKeywordWithBlankPayload
display - TestDisplayKeywordWithGuestUser
upload - TestUploadKeywordWithGuestUser
display - TestDisplayKeywordWithAuthenticatedUser
```

I am not sure now if `TestUploadKeywordWithValidParams` and `TestUploadKeywordWithBlankPayload` are for guest or authenticated users unless I check carefully the whole files.

It's more standard to group tests by:
- Order of appearance of the methods in the file under tests. Since the simply method appears first, then all tests for display should be first. Then the tests for upload can come after that.
- Happy path before sad path: so tests authenticated user should come before tests with no authenticated users within the group of tests for each method.

Reply:
Good catch Fixed in 423332f7a1b2b4fa1a7f52edd71809c496ddafaa.

Maybe should this conversation be a part of our Go convention?

Reply:
This seems more like general best practice for all languages but if you feel it's needed to specify for Go then go for it.